homeassistant/helpers/entity.py

```diff
@@ -259,7 +261,130 @@ class CalculatedState:
     shadowed_attributes: Mapping[str, Any]
 
 
-class Entity(ABC):
+class CachedProperties(type):
+    """Metaclass which invalidates cached entity propertis on write to _attr_."""
+
+    def __new__(
+        mcs,  # noqa: N804  ruff bug, ruff does not understand this is a metaclass
+        name: str,
+        bases: tuple[type, ...],
+        namespace: dict[Any, Any],
+        cached_properties: set[str] | None = None,
+        **kwargs: Any,
+    ) -> Any:
+        """Pop cached_properties and store it in the namespace."""
+        namespace["_CachedProperties__cached_properties"] = cached_properties or set()
+        return super().__new__(mcs, name, bases, namespace)
+
+    def __init__(
+        cls,
+        name: str,
+        bases: tuple[type, ...],
+        namespace: dict[Any, Any],
+        **kwargs: Any,
+    ) -> None:
+        """Wrap _attr_ for cached properties in property objects."""
+
+        def deleter(name: str) -> Callable[[Any], None]:
+            private_attribute_name = "__attr_" + name
+
+            def _deleter(o: Any) -> None:
+                for attr in (name, private_attribute_name):
+                    try:  # noqa: SIM105  suppress is much slower
+                        delattr(o, attr)
+                    except AttributeError:
+                        pass
+
+            return _deleter
+
+        def getter(name: str) -> Callable[[Any], Any]:
+            private_attribute_name = "__attr_" + name
+
+            def _getter(o: Any) -> Any:
+                return getattr(o, private_attribute_name)
+
+            return _getter
+
+        def setter(name: str) -> Callable[[Any, Any], None]:
+            private_attribute_name = "__attr_" + name
+
+            def _setter(o: Any, val: Any) -> None:
+                setattr(o, private_attribute_name, val)
+                try:  # noqa: SIM105  suppress is much slower
+                    delattr(o, name)
+                except AttributeError:
+                    pass
+
+            return _setter
+
+        def make_property(name: str) -> property:
+            return property(fget=getter(name), fset=setter(name), fdel=deleter(name))
+
+        def move_attr(cls: CachedProperties, property_name: str) -> None:
+            attr_name = "_attr_" + property_name
+            private_attr_name = "__attr_" + property_name
```

Comment:
Are f-strings slower than string concatenation?

Reply:
Should be fine to use f-strings here

Reply:
It doesn't matter, this runs once per class. Anyhow, comments are needed, thanks for pushing for it @MartinHjelmare 👍